lenskit-core/src/main/java/org/lenskit/rerank/GreedyRerankingItemRecommender.java

```diff
@@ -0,0 +1,77 @@
+package org.lenskit.rerank;
+
+import it.unimi.dsi.fastutil.longs.LongSet;
+import org.lenskit.api.ItemRecommender;
+import org.lenskit.api.Result;
+import org.lenskit.api.ResultList;
+import org.lenskit.basic.AbstractItemRecommender;
+import org.lenskit.results.Results;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * A hybrid item recommender that uses a greedy re-ranking strategy to allow re-ranking of items iteratively. This
+ * general algorithm is commonly employed to efficiently optimize for set properties of a recommendation list such as
+ * inter-item diveristy.
+ *
+ * This algorithm takes a baseline ranking algorithm, gets the top-n recommendations and re-ranks them iteratively.
+ * To select each recommended item, first a scoring algorithm is ran based on the currently selected recommendations
+ * and each candidate item. The item with the highest score is then added to the recommended list. This process repeates
+ * until enough items are recommended.
+ *
+ * @author Daniel Kluver
+ */
+public class GreedyRerankingItemRecommender extends AbstractItemRecommender {
+    private static final Logger logger = LoggerFactory.getLogger(GreedyRerankingItemRecommender.class);
+    private final ItemRecommender baseRecommender;
+    private final int numCandidates;
+    private final Rescorer rescorer;
+
+    @Inject
+    public GreedyRerankingItemRecommender(ItemRecommender baseRecommender, @NumCandidates int numCandidates, Rescorer rescorer) {
+        this.baseRecommender = baseRecommender;
+        this.numCandidates = numCandidates;
+        this.rescorer = rescorer;
+    }
+
+
+    @Override
+    protected ResultList recommendWithDetails(long user, int n, @Nullable LongSet candidateItems, @Nullable LongSet exclude) {
+        List<Result> candidates = baseRecommender.recommendWithDetails(user, numCandidates, candidateItems, exclude);
+        //TODO: is there a better option for this?
+        //modifiable copy
+        candidates = new ArrayList<>(candidates);
+
+        if (n<0) {
+            n = candidates.size();
+        }
+
+        //TODO: use ResultLists instead of Lists
+        //TODO: is this the best List for the situation?
+
+        List<GreedyRerankingResult> results = new ArrayList<>(n);
+        for (int i = 0; i<n; i++) {
+            // TODO: I would like to make this a private function for better method decomposition, but then I have to recompute scores.
+            Result bestResult = null;
+            double bestScore = 0;
+            for (Result option : candidates) {
```

Comment:
Seems like we have an interesting performance issue here.

For the general case, looking at score is good. And we clearly need to scan to find 'best', but it might be helpful to have a 'but you won't find better' signal.

Or maybe the right abstraction boundary is to ask the inner to pick instead of score. That way it can pick more efficiently: for a binary case, it can just use the first one that meets the binary condition.